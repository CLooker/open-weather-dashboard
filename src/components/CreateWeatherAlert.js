import React, { Component } from 'react';
import apiKey from '../api';
import PropTypes from 'prop-types';
import moment from 'moment';

export default class CreateWeatherAlert extends Component {
  static propTypes = {
    incrementRegisteredAlerts: PropTypes.func.isRequired
  };

  state = {
    query: '',
    units: '',
    message: ''
  };

  getBody = ({ start, query, units }) => ({
    time_period: {
      start: {
        expression: 'after',
        amount: start
      },
      end: {
        expression: 'after',
        amount: moment(start)
          .add(1, 'week')
          .valueOf()
      }
    },
    conditions: [
      {
        name: 'temp',
        expression: '$gt',
        amount: this.handleConversion({ query, units })
      }
    ],
    area: [
      {
        type: 'Point',
        coordinates: [41.9, 12.5]
      }
    ]
  });

  handleInputChange = e =>
    this.setState({
      query: e.target.value
    });

  handleUnitsChange = e =>
    this.setState({
      units: e.target.value
    });

  handleSubmit = e => {
    const { query, units } = this.state;
    e.preventDefault();
    if (this.validate(query, units)) {
      this.setState({
        message: ''
      });
      fetch(
        `http://api.openweathermap.org/data/3.0/triggers?&APPID=${apiKey}`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify(
            this.getBody({ query, units, start: moment().valueOf() })
          )
        }
      )
        .then(res => res.json())
        .then(res => this.afterFormSubmit({ type: 'success' }))
        .catch(err => {
          console.log(err);
          this.afterFormSubmit({ type: 'failure' });
        });
    }
  };

  afterFormSubmit = ({ type }) => {
    type === 'success'
      ? this.setState(prevState => ({
          query: '',
          units: '',
          message: 'Success.'
        })) || this.props.incrementRegisteredAlerts()
      : this.setState({
          query: '',
          units: '',
          message: 'Something went wrong.'
        });
  };

  validate = (temp, units) =>
    this.validateInput(temp) && this.validateUnits(units);

  validateInput = input =>
    !input || isNaN(input)
      ? alert('Please enter a number for temperature.') || false
      : true;

  validateUnits = units =>
    !units ? alert('Please select units.') || false : true;

  handleConversion = ({ query, units }) => {
    switch (units) {
      case 'C':
        return parseInt(query, 10) + 273.15;
      case 'F':
        return (parseInt(query, 10) + 459.67) * (5 / 9);
      default:
        return parseInt(query, 10);
    }
  };

  render() {
    const { query, units, message } = this.state;
    return (
      <div className="create-weather-alert-container">
        <div className="create-weather-alert">
          <div className="create-weather-alert-title-icon">
            <img
              src={`https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQYVYTeX-IT8VF_M7yQQsClU2CQEJTXgoi8T9jaGhB66jpLWQSw`}
              alt="Icon depicting current weather"
            />
            <p>Create Max Temperature Alert</p>
          </div>
          <form onSubmit={this.handleSubmit}>
            <div>
              <input
                type="text"
                onChange={this.handleInputChange}
                value={query}
              />
              <select onChange={this.handleUnitsChange} value={units}>
                <option value="" />
                <option value="F">&deg;F</option>
                <option value="C">&deg;C</option>
                <option value="K">&deg;K</option>
              </select>
            </div>
            <input type="submit" value="Submit" />
          </form>
          <div>
            <h5>{message}</h5>
          </div>
        </div>
      </div>
    );
  }
}
