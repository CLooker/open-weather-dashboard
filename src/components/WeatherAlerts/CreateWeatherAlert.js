import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import moment from 'moment';
import { apiKey } from '../../utils';
import { IconAndText } from '../common';

const CreateWeatherAlertWrapper = styled.div`
  margin: 0 15px;
`;

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  flex-direction: column;
`;

const IconAndTextWrapper = styled(IconAndText)`
  margin: 5px 0 10px 0;
  display: flex;
`;

const SubmitWrapper = styled.div`
  display: flex;
  justify-content: center;
  margin: 5px 0 0;
`;

export default class CreateWeatherAlert extends Component {
  static propTypes = {
    incrementRegisteredAlertsTotal: PropTypes.func.isRequired
  };

  state = {
    alertTemperature: '',
    displayMessage: '',
    units: ''
  };

  setAlertTemperature = e =>
    this.setState({
      alertTemperature: e.target.value
    });

  setUnits = e =>
    this.setState({
      units: e.target.value
    });

  getReqBody = () => {
    const now = moment().valueOf();

    return JSON.stringify({
      // from now until 1 week
      time_period: {
        start: {
          expression: 'after',
          amount: now
        },
        end: {
          expression: 'before',
          amount: moment(now)
            .add(1, 'week')
            .valueOf()
        }
      },
      conditions: [
        {
          name: 'temp',
          expression: '$gt',
          amount: this.handleTempConversion()
        }
      ],
      area: [
        {
          type: 'Point',
          coordinates: [41.9, 12.5]
        }
      ]
    });
  };

  handleTempConversion = () => {
    const { alertTemperature, units } = this.state;
    switch (units) {
      case 'C':
        return +alertTemperature + 273.15;
      case 'F':
        return (+alertTemperature + 459.67) * (5 / 9);
      default:
        return +alertTemperature;
    }
  };

  handleSubmit = async e => {
    e.preventDefault();

    if (!this.isValidInput()) return;

    await this.setState({ displayMessage: '' });

    fetch(`https://api.openweathermap.org/data/3.0/triggers?&APPID=${apiKey}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: this.getReqBody()
    })
      .then(res => res.json())
      .then(this.handleSubmitSuccess)
      .catch(this.handleSubmitFailure);
  };

  isValidInput = () => {
    const { alertTemperature, units } = this.state;

    const isValidTemp = (() => {
      if (alertTemperature && Number.isInteger(+alertTemperature)) return true;

      alert('Please enter an integer for creating a weather alert.');
      return false;
    })();
    if (!isValidTemp) return false;

    const isValidUnits = (() => {
      if (units) return true;

      alert('Please select temperature units for creating a weather alert.');
      return false;
    })();

    return isValidUnits;
  };

  handleSubmitSuccess = async () => {
    const { incrementRegisteredAlerts } = this.props;

    await this.setState({ displayMessage: 'Success.' });
    incrementRegisteredAlerts();

    // make Success message disappear in 10 seconds
    setTimeout(
      () =>
        this.setState({ alertTemperature: '', displayMessage: '', units: '' }),
      10000
    );
  };

  handleSubmitFailure = () =>
    this.setState({ displayMessage: 'Something went wrong.' });

  render() {
    const { alertTemperature, displayMessage, units } = this.state;

    return (
      <CreateWeatherAlertWrapper>
        <Wrapper>
          <IconAndTextWrapper
            iconStyle={{ width: '25px', height: '25px' }}
            titleAttr={'create alert'}
            iconSrc='https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQYVYTeX-IT8VF_M7yQQsClU2CQEJTXgoi8T9jaGhB66jpLWQSw'
            text={'Create Max Temperature Alert'}
          />
          <form onSubmit={this.handleSubmit}>
            <div>
              <input
                type='text'
                onChange={this.setAlertTemperature}
                value={alertTemperature}
              />
              <select onChange={this.setUnits} value={units}>
                <option value='' />
                <option value='F'>&deg;F</option>
                <option value='C'>&deg;C</option>
                <option value='K'>&deg;K</option>
              </select>
            </div>
            <SubmitWrapper>
              <input type='submit' value='Submit' />
            </SubmitWrapper>
          </form>
          <div>
            <h5>{displayMessage}</h5>
          </div>
        </Wrapper>
      </CreateWeatherAlertWrapper>
    );
  }
}
