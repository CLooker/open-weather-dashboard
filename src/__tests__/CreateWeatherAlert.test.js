import React from 'react';
import { mount, shallow } from 'enzyme';
import CreateWeatherAlert from '../components/CreateWeatherAlert';
import moment from 'moment';

describe('CreateWeatherAlert', () => {
  const props = {
    incrementRegisteredAlerts: jest.fn()
  };

  let component;

  beforeEach(() => (component = shallow(<CreateWeatherAlert {...props} />)));

  afterEach(() => jest.restoreAllMocks());

  describe('initialization', () => {
    it('should render correctly', () => {
      expect(component).toMatchSnapshot();
    });

    it('should have certain props', () => {
      const actual = component.instance().props;
      const expected = props;
      expect(actual).toEqual(props);
    });

    it('should have certain state', () => {
      const actual = component.state();
      const expected = {
        query: '',
        units: '',
        message: ''
      };
      expect(actual).toEqual(expected);
    });

    describe('rendering', () => {
      const mountedComponent = mount(<CreateWeatherAlert {...props} />);

      it('should render state', () => {
        const state = {
          query: '100',
          units: 'F',
          message: 'Success.'
        };
        mountedComponent.setState(state);

        const actual = mountedComponent.html();
        Object.keys(state).forEach(key => {
          expect(actual).toContain(state[key]);
        });
      });
    });
  });

  describe('methods', () => {
    describe('`getBody`', () => {
      it('should return an object of a certain shape', () => {
        const query = 300;
        const units = 'K';
        const actual = component
          .instance()
          .getBody({ query, units, start: moment().valueOf() });
        const expected = {
          time_period: {
            start: {
              expression: 'after',
              amount: expect.any(Number)
            },
            end: {
              expression: 'after',
              amount: expect.any(Number)
            }
          },
          conditions: [
            {
              name: 'temp',
              expression: '$gt',
              amount: parseInt(query, 10)
            }
          ],
          area: [
            {
              type: 'Point',
              coordinates: [41.9, 12.5]
            }
          ]
        };
        expect(actual).toEqual(expected);
      });
    });

    describe('`handleInputChange`', () => {
      it('should setState correctly', () => {
        const target = {
          value: '100'
        };
        component.instance().handleInputChange({ target });

        const actual = component.state().query;
        const expected = target.value;
        expect(actual).toBe(expected);
      });
    });

    describe('`handleUnitsChange`', () => {
      beforeEach(() => {});

      it('should setState correctly', () => {
        const target = {
          value: 'C'
        };
        component.instance().handleUnitsChange({ target });

        const actual = component.state().units;
        const expected = target.value;
        expect(actual).toBe(expected);
      });
    });

    //TODO
    describe('`handleSubmit`', () => {});

    describe('`validate`', () => {
      it('should call `validateInput`', () => {
        const spy = jest.spyOn(component.instance(), 'validate');
        component.instance().validate(100, 'K');
        expect(spy).toHaveBeenCalled();
      });
      describe('code paths', () => {
        describe('`validateInput` returns `true`', () => {
          it('should call `validateUnits`', () => {
            const spy = jest.spyOn(component.instance(), 'validateUnits');
            component.instance().validateInput = jest.fn(() => true);
            component.instance().validate(100, 'K');
            expect(spy).toHaveBeenCalled();
          });

          it('should return what `validateUnits` returns', () => {
            const returnValue = false;
            component.instance().validateInput = jest.fn(() => true);
            component.instance().validateUnits = jest.fn(() => returnValue);
            const actual = component.instance().validate(100, 'K');
            const expected = returnValue;
            expect(actual).toBe(expected);
          });
        });

        describe('`validateInput` returns `false`', () => {
          it('should not call `validateUnits`', () => {
            const spy = jest.spyOn(component.instance(), 'validateUnits');
            component.instance().validateInput = jest.fn(() => false);
            component.instance().validate(100, 'K');
            expect(spy).not.toHaveBeenCalled();
          });

          it('should return what `validateInput` returns', () => {
            const returnValue = false;
            component.instance().validateInput = jest.fn(() => returnValue);
            const actual = component.instance().validate(100, 'K');
            const expected = returnValue;
            expect(actual).toBe(expected);
          });
        });
      });
    });

    describe('`validateInput`', () => {
      describe('code paths', () => {
        describe('no input is given', () => {
          it('should call `alert`', () => {
            global.alert = jest.fn();
            component.instance().validateInput();
            expect(global.alert).toHaveBeenCalled();
          });

          it('should return `false`', () => {
            const actual = component.instance().validateInput();
            const expected = false;
            expect(actual).toBe(expected);
          });
        });

        describe('input is given', () => {
          describe('input is `NaN`', () => {
            const input = 'a string';

            it('should call `alert`', () => {
              global.alert = jest.fn();
              component.instance().validateInput(input);
              expect(global.alert).toHaveBeenCalled();
            });

            it('should return `false`', () => {
              const actual = component.instance().validateInput(input);
              const expected = false;
              expect(actual).toBe(expected);
            });
          });

          describe('input is a number', () => {
            it('should return true', () => {
              const input = 100;
              const actual = component.instance().validateInput(input);
              const expected = true;
              expect(actual).toBe(expected);
            });
          });
        });
      });
    });

    describe('`validateUnits`', () => {
      describe('code paths', () => {
        describe('no units are given', () => {
          it('should call `alert`', () => {
            global.alert = jest.fn();
            component.instance().validateUnits();
            expect(global.alert).toHaveBeenCalled();
          });

          it('should return `false`', () => {
            const actual = component.instance().validateUnits();
            const expected = false;
            expect(actual).toBe(expected);
          });
        });

        describe('units are given', () => {
          it('should return `true`', () => {
            const units = 'K';
            const actual = component.instance().validateUnits(units);
            const expected = true;
            expect(actual).toBe(expected);
          });
        });
      });
    });

    describe('`handleConversion`', () => {
      describe('code paths', () => {
        describe('`units`', () => {
          describe('"C"', () => {
            it('should return query converted to celsius', () => {
              const query = 100;
              const units = 'C';
              const actual = component
                .instance()
                .handleConversion({ query, units });
              const expected = parseInt(query, 10) + 273.15;
              expect(actual).toBe(expected);
            });
          });

          describe('"F"', () => {
            it('should return query converted to fahrenheit', () => {
              const query = 100;
              const units = 'F';
              const actual = component
                .instance()
                .handleConversion({ query, units });
              const expected = (parseInt(query, 10) + 459.67) * (5 / 9);
              expect(actual).toBe(expected);
            });
          });

          describe('"K"', () => {
            it('should return query', () => {
              const query = 100;
              const units = 'K';
              const actual = component
                .instance()
                .handleConversion({ query, units });
              const expected = query;
              expect(actual).toBe(expected);
            });
          });
        });
      });
    });

    describe('`afterFormSubmit`', () => {
      describe('`type` is "success"', () => {
        it('should `setState` correctly', () => {
          component.instance().afterFormSubmit({ type: 'success' });
          const expected = {
            query: '',
            units: '',
            message: 'Success.'
          };
          const actual = component.state();
          expect(actual).toEqual(expect.objectContaining(expected));
        });

        it('should call `incrementRegisteredAlerts`', () => {
          const mountedComponent = mount(<CreateWeatherAlert {...props} />);
          component.instance().afterFormSubmit({ type: 'success' });
          expect(
            mountedComponent.props().incrementRegisteredAlerts
          ).toHaveBeenCalled();
        });
      });

      describe('`type` is "failure"', () => {
        it('should `setState` correctly', () => {
          component.instance().afterFormSubmit({ type: 'failure' });
          const expected = {
            query: '',
            units: '',
            message: 'Something went wrong.'
          };
          const actual = component.state();
          expect(actual).toEqual(expect.objectContaining(expected));
        });
      });
    });
  });

  //TODO
  describe('events', () => {});
});
