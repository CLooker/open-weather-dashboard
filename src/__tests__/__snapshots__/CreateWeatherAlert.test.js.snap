// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateWeatherAlert initialization should render correctly 1`] = `
"<div className=\\"create-weather-alert-container\\">
  <div className=\\"create-weather-alert\\">
    <div className=\\"create-weather-alert-title-icon\\">
      <img src=\\"https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQYVYTeX-IT8VF_M7yQQsClU2CQEJTXgoi8T9jaGhB66jpLWQSw\\" alt=\\"Icon depicting current weather\\" />
      <p>
        Create Max Temperature Alert
      </p>
    </div>
    <form onSubmit={[Function]}>
      <div>
        <input type=\\"text\\" onChange={[Function]} value=\\"\\" />
        <select onChange={[Function]} value=\\"\\">
          <option value=\\"\\" />
          <option value=\\"F\\">
            °F
          </option>
          <option value=\\"C\\">
            °C
          </option>
          <option value=\\"K\\">
            °K
          </option>
        </select>
      </div>
      <input type=\\"submit\\" value=\\"Submit\\" />
    </form>
    <div>
      <h5 />
    </div>
  </div>
</div>"
`;
